token/src/token.rs

```diff
@@ -0,0 +1,89 @@
+use bulletproofs::{BulletproofGens, PedersenGens};
+use curve25519_dalek::scalar::Scalar;
+use zkvm::*;
+
+pub struct Token {
+    flavor: Predicate,
+    metadata: Vec<u8>,
+    nonce: Scalar,
+}
+
+impl Token {
+    pub fn new(pred: Predicate, metadata: &[u8]) -> Self {
+        Token {
+            flavor: pred,
+            metadata: metadata.to_vec(),
+            nonce: Scalar::random(&mut rand::thread_rng()),
+        }
+    }
+
+    pub fn issue<'a>(&self, program: &'a mut Program, qty: u64) -> &'a mut Program {
+        let nonce = Predicate::Key(VerificationKey::from_secret(&self.nonce));
+        program
+            .push(Commitment::blinded(qty)) // stack: qty
+            .var() // stack: qty-var
+            .push(Commitment::unblinded(Value::issue_flavor(
+                &self.flavor,
+                Data::Opaque(self.metadata.to_vec()),
+            ))) // stack: qty-var, flv
+            .var() // stack: qty-var, flv-var
+            .push(Data::Opaque(self.metadata.to_vec())) // stack: qty-var, flv-var, data
+            .push(self.flavor.clone()) // stack: qty-var, flv-var, data, flv-pred
+            .issue() // stack: issue-contract
+            .push(nonce) // stack: issue-contract, nonce-pred
+            .nonce() // stack: issue-contract, nonce-contract
+            .sign_tx() // stack: issue-contract
+            .sign_tx() // stack: issued-value
+    }
+
+    pub fn issue_and_spend<'a>(
+        &self,
+        program: &'a mut Program,
+        qty: u64,
+        dest: Predicate,
+    ) -> &'a mut Program {
+        self.issue(program, qty).push(dest).output(1)
+    }
+
+    pub fn build(
```

Comment:
maybe keep it as a helper method in the tests - it's not likely that the `Token` API would take care of an entire transaction - that may be handled in a "tx builder" API or simply in a concrete application that assembles transactions from various pieces.